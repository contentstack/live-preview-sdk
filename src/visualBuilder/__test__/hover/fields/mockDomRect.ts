export const mockDomRect = {
    singleHorizontal: () => ({
        bottom: 88.3984375,
        height: 54.3984375,
        left: 34,
        right: 862,
        top: 34,
        width: 828,
        x: 34,
        y: 34,
    }),
    singleLeft: () => ({
        x: 51,
        y: 51,
        width: 27.7734375,
        height: 20.3984375,
        top: 51,
        bottom: 71.3984375,
        left: 51,
        right: 78.7734375,
    }),
    singleRight: () => ({
        x: 110.7734375,
        y: 51,
        width: 53.59375,
        height: 20.3984375,
        top: 51,
        bottom: 71.3984375,
        left: 110.7734375,
        right: 164.3671875,
    }),
    singleVertical: () => ({
        bottom: 195.1953125,
        height: 90.796875,
        left: 34,
        right: 862,
        top: 104.3984375,
        width: 828,
        x: 34,
        y: 104.3984375,
    }),
    singleTop: () => ({
        x: 51,
        y: 121.3984375,
        width: 794,
        height: 20.3984375,
        top: 121.3984375,
        bottom: 141.796875,
        left: 51,
        right: 845,
    }),
    singleBottom: () => ({
        x: 51,
        y: 157.796875,
        width: 794,
        height: 20.3984375,
        top: 157.796875,
        bottom: 178.1953125,
        left: 51,
        right: 845,
    }),
    multipleLeft: () => ({
        x: 51,
        y: 228.1953125,
        width: 87.59375,
        height: 90.796875,
        top: 228.1953125,
        bottom: 318.9921875,
        left: 51,
        right: 138.59375,
    }),
    multipleRight: () => ({
        x: 170.59375,
        y: 228.1953125,
        width: 87.59375,
        height: 90.796875,
        top: 228.1953125,
        bottom: 318.9921875,
        left: 170.59375,
        right: 258.1875,
    }),
    multipleTop: () => ({
        x: 51,
        y: 368.9921875,
        width: 794,
        height: 90.796875,
        top: 368.9921875,
        bottom: 459.7890625,
        left: 51,
        right: 845,
    }),
    multipleBottom: () => ({
        x: 51,
        y: 475.7890625,
        width: 794,
        height: 90.796875,
        top: 475.7890625,
        bottom: 566.5859375,
        left: 51,
        right: 845,
    }),
    multipleChildTop: () => ({
        x: 68,
        y: 385.9921875,
        width: 760,
        height: 20.3984375,
        top: 385.9921875,
        bottom: 406.390625,
        left: 68,
        right: 828,
    }),
    multipleChildBottom: () => ({
        x: 68,
        y: 422.390625,
        width: 760,
        height: 20.3984375,
        top: 422.390625,
        bottom: 442.7890625,
        left: 68,
        right: 828,
    }),
};
