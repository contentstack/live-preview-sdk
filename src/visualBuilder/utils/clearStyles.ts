const originalStyles = new WeakMap<
    HTMLElement,
    {
        visibility?: string;
        transition?: string;
        animation?: string;
    }
>();

export function clearVisibelityStyles(element: HTMLElement) {
    const originalStyleValues = {
        visibility: element.style.visibility,
        transition: element.style.transition,
        animation: element.style.animation,
    };

    originalStyles.set(element, originalStyleValues);

    element.style.visibility = "hidden";
    element.style.transition = "none";
    element.style.animation = "none";
}

export function restoreVisibelityStyles(element: HTMLElement) {
    const storedStyles = originalStyles.get(element);

    if (storedStyles) {
        element.style.visibility = storedStyles.visibility || "";
        element.style.transition = storedStyles.transition || "";
        element.style.animation = storedStyles.animation || "";

        originalStyles.delete(element);
    }
}
