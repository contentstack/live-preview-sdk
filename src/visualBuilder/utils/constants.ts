import { FieldDataType } from "./types/index.types";

export const numericInputRegex = /^-?\d*(\.\d*)?([eE][-+]?\d*)?$/;

export const VISUAL_BUILDER_FIELD_TYPE_ATTRIBUTE_KEY = "data-cslp-field-type";

export const VISUAL_BUILDER_CHANNEL_ID = "visual-builder";

export const LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX = 2;

// Minimum distance from top edge to prevent toolbar from being hidden
export const TOP_EDGE_BUFFER = 42;
export const RIGHT_EDGE_BUFFER = 180;

export const TOOLBAR_EDGE_BUFFER = 8;

export const DATA_CSLP_ATTR_SELECTOR = "data-cslp";

/**
 * The field that can be directly modified using contenteditable=true.
 * This includes all text fields like title and numbers.
 */
export const ALLOWED_INLINE_EDITABLE_FIELD: FieldDataType[] = [
    FieldDataType.SINGLELINE,
    FieldDataType.MULTILINE,
    FieldDataType.NUMBER,
];

export const ALLOWED_MODAL_EDITABLE_FIELD: FieldDataType[] = [
    FieldDataType.HTML_RTE,
    FieldDataType.MARKDOWN_RTE,
    FieldDataType.JSON_RTE,
    FieldDataType.CUSTOM_FIELD,
    FieldDataType.LINK,
    FieldDataType.ISODATE,
    FieldDataType.URL,
];

export const ALLOWED_REPLACE_FIELDS: FieldDataType[] = [
    FieldDataType.REFERENCE,
    FieldDataType.FILE,
];

export const DEFAULT_MULTIPLE_FIELDS: FieldDataType[] = [
    FieldDataType.GLOBAL_FIELD,
    FieldDataType.GROUP,
    FieldDataType.BLOCK,
];

export const mentionLimit = 20;
export const maxMessageLength = 500;
