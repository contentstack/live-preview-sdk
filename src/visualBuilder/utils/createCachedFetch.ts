/**
 * Creates a cached async fetch function with support for any number of arguments
 * @param fetchFn - The async function to cache
 * @param uidResolver - Function that generates a unique ID from the arguments passed to fetchFn
 * @returns A cached version of the fetch function with the same signature
 */
export function createCachedFetch<TArgs extends any[], TResult>(
    fetchFn: (...args: TArgs) => Promise<TResult>,
    uidResolver: (...args: TArgs) => string = (...args) => JSON.stringify(args)
): {
    (...args: TArgs): Promise<TResult>;
    clearCache: () => void;
} {
    // Cache storage
    const cache = new Map<string, TResult>();
    // Track in-flight requests
    const pendingPromises = new Map<string, Promise<TResult>>();

    /**
     * The cached fetch function
     * @param args - Arguments to pass to the original fetch function
     * @returns Promise that resolves with the data
     */
    async function cachedFetch(...args: TArgs): Promise<TResult> {
        // Generate unique ID for these arguments
        const uid = uidResolver(...args);

        // Return cached value if available
        if (cache.has(uid)) {
            return cache.get(uid)!;
        }

        // Return existing promise if request is already in progress
        if (pendingPromises.has(uid)) {
            return pendingPromises.get(uid)!;
        }

        // Create new promise for this request
        const promise = fetchFn(...args)
            .then((data) => {
                // Store result in cache
                cache.set(uid, data);
                // Remove from pending
                pendingPromises.delete(uid);
                return data;
            })
            .catch((error) => {
                // Clean up on error
                pendingPromises.delete(uid);
                throw error;
            });

        // Store the promise
        pendingPromises.set(uid, promise);
        return promise;
    }

    // Add clearCache method to the function
    cachedFetch.clearCache = () => {
        cache.clear();
        pendingPromises.clear();
    };

    return cachedFetch;
}
