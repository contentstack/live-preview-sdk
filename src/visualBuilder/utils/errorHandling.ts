
export function hasPostMessageError(obj: any): boolean {
    return obj?.error === true;
}