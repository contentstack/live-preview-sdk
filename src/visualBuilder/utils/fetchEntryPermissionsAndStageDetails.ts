import { getEntryPermissionsCached } from "./getEntryPermissionsCached";
import { getWorkflowStageDetails } from "./getWorkflowStageDetails";

export async function fetchEntryPermissionsAndStageDetails({
    entryUid,
    contentTypeUid,
    locale,
    variantUid,
}: {
    entryUid: string;
    contentTypeUid: string;
    locale: string;
    variantUid?: string | undefined;
}) {
    const entryAclPromise = getEntryPermissionsCached({
        entryUid,
        contentTypeUid,
        locale,
    });
    const entryWorkflowStageDetailsPromise = getWorkflowStageDetails({
        entryUid,
        contentTypeUid,
        locale,
        variantUid,
    });
    const results = await Promise.allSettled([
        entryAclPromise,
        entryWorkflowStageDetailsPromise,
    ]);
    if (results[0].status === "rejected") {
        console.debug(
            "[Visual Builder] Error retrieving entry permissions",
            results[0].reason
        );
    }
    if (results[1].status === "rejected") {
        console.debug(
            "[Visual Builder] Error retrieving entry stage details",
            results[1].reason
        );
    }
    const acl =
        results[0].status === "fulfilled" ? results[0].value : undefined;
    const workflowStage =
        results[1].status === "fulfilled" ? results[1].value : undefined;
    return {
        acl,
        workflowStage,
    };
}
