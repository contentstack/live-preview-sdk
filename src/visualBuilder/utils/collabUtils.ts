import {
    ICommentState,
    IMentionedList,
    IMentionItem,
    IMessageDTO,
    IUserDTO,
    IUserState,
} from "../types/collab.types";
import { maxMessageLength, mentionLimit } from "./constants";
import { uniqBy } from "lodash-es";
import DOMPurify from "dompurify";

const escapeRegExp = (string: string): string => {
    return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
};

/**
 * Generates the title for the thread based on the number of comments.
 * @param {number} commentCount - The number of comments.
 * @returns {string} The title for the thread.
 */
export const getThreadTitle = (commentCount: number): string => {
    if (commentCount === 0) return "Add New Feedback";
    return commentCount === 1 ? "1 Feedback" : `${commentCount} Feedbacks`;
};

/**
 * returns the available email.
 * @param {IUserDTO} user - The user object.
 * @returns {string} The user's email.
 */
export const getUserName = (user: IUserDTO): string => {
    return user.email;
};

/**
 * Validates the comment length and the number of mentions.
 * @param {string} comment - The comment message.
 * @param {IMentionedList} toUsers - The list of mentioned users.
 * @returns {string} The error message if validation fails, otherwise an empty string.
 */
export const validateCommentAndMentions = (
    comment: string,
    toUsers: IMentionedList
): string => {
    if (comment.length > maxMessageLength) {
        return `Limit exceeded. You can have a maximum length of ${maxMessageLength} characters.`;
    }
    if (toUsers.length > mentionLimit) {
        return `Limit exceeded. You can tag a maximum of ${mentionLimit} users.`;
    }
    return "";
};

/**
 * Removes mentions that no longer exist in the message.
 * @param {string} message - The comment message.
 * @param {IMentionedList} toUsers - The list of mentioned users.
 * @returns {Object} The updated lists of mentioned users.
 */
export const filterOutInvalidMentions = (
    message: string,
    toUsers: IMentionedList
) => {
    const to_users_temp = toUsers.filter((user) =>
        message.includes(user.display)
    );

    return {
        toUsers: uniqBy(to_users_temp, "id"),
    };
};

/**
 * Replaces mention placeholders with display names in the comment message.
 * @param {IMessageDTO | undefined} comment - The comment object.
 * @param {IUserState} userState - The user state containing user and role maps.
 * @param {"text" | "html"} profile - The format for the output message, either plain text or HTML.
 * @returns {string | undefined} The formatted message or undefined if the comment is not provided.
 */
export const getMessageWithDisplayName = (
    comment: IMessageDTO | undefined | null,
    userState: IUserState,
    profile: "text" | "html"
): string | undefined => {
    if (!comment) return undefined;

    let tempText = comment.message;

    comment?.toUsers?.forEach((user) => {
        const userPattern = new RegExp(`{{${user}}}`, "g");
        const userData = userState.userMap[user];
        const replacement =
            profile === "html"
                ? `<b>@${userData.display || getUserName(userData)}</b>`
                : `@${userData.display || getUserName(userData)}`;
        tempText = tempText.replace(userPattern, replacement);
    });

    return tempText;
};

/**
 * Sanitizes HTML content to prevent XSS attacks.
 * @param {any} dirty - The unsanitized HTML content.
 * @returns {string} The sanitized HTML content.
 */
export const sanitizeData = (dirty: any): string => {
    return DOMPurify.sanitize(dirty, { USE_PROFILES: { html: true } });
};

/**
 * Constructs the comment body with mentions replaced by their unique identifiers.
 * @param {ICommentState} state - The state containing the comment and mentions.
 * @returns {Object} The comment body containing the sanitized message and mentioned users.
 */
export const getCommentBody = (state: ICommentState): ICommentState => {
    let finalMessage = state.message.trim().replace(/\s+/g, " ");
    const comment = {
        message: finalMessage,
        toUsers: [],
        images: [],
        createdBy: state.createdBy,
        author: state.author,
    };

    const updateMentionToUID = (
        entity: IMentionItem,
        result: Array<string>
    ) => {
        const displayName = entity.display;

        const escapedDisplayName = escapeRegExp(displayName);
        const regexUser = new RegExp(escapedDisplayName, "g");
        finalMessage = finalMessage.replace(regexUser, `{{${entity.id}}}`);
        result.push(entity.display);
    };

    state.toUsers?.forEach((user) => updateMentionToUID(user, comment.toUsers));

    comment.message = finalMessage;
    return comment;
};
