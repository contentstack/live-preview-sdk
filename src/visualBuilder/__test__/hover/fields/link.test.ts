import { screen } from "@testing-library/preact";
import { getFieldSchemaMap } from "../../../../__test__/data/fieldSchemaMap";
import { waitForHoverOutline } from "../../../../__test__/utils";
import Config from "../../../../configManager/configManager";
import { VisualBuilder } from "../../../index";
import { FieldSchemaMap } from "../../../utils/fieldSchemaMap";
import { mockDomRect } from "./mockDomRect";

vi.mock("../../../utils/visualBuilderPostMessage", async () => {
    const { getAllContentTypes } = await vi.importActual<
        typeof import("../../../../__test__/data/contentType")
    >("../../../../__test__/data/contentType");
    const contentTypes = getAllContentTypes();
    return {
        __esModule: true,
        default: {
            send: vi.fn().mockImplementation((eventName: string) => {
                if (eventName === "init")
                    return Promise.resolve({
                        contentTypes,
                    });
                return Promise.resolve();
            }),
        },
    };
});

global.ResizeObserver = vi.fn().mockImplementation(() => ({
    observe: vi.fn(),
    unobserve: vi.fn(),
    disconnect: vi.fn(),
}));

describe("When an element is hovered in visual builder mode", () => {
    let mousemoveEvent: Event;

    beforeAll(() => {
        FieldSchemaMap.setFieldSchema(
            "all_fields",
            getFieldSchemaMap().all_fields
        );
    });

    beforeEach(() => {
        Config.reset();
        Config.set("mode", 2);
        mousemoveEvent = new Event("mousemove", {
            bubbles: true,
            cancelable: true,
        });
    });

    afterEach(() => {
        vi.clearAllMocks();
        document.getElementsByTagName("html")[0].innerHTML = "";
    });

    afterAll(() => {
        Config.reset();
    });

    describe("link field", () => {
        let linkField: HTMLAnchorElement;
        let visualBuilder: VisualBuilder;

        beforeEach(() => {
            linkField = document.createElement("a");
            linkField.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.link.href"
            );

            linkField.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleLeft());

            document.body.appendChild(linkField);

            visualBuilder = new VisualBuilder();
        });

        afterEach(() => {
            visualBuilder.destroy();
        });

        test("should have outline and custom cursor", async () => {
            linkField.dispatchEvent(mousemoveEvent);
            await waitForHoverOutline();
            const hoverOutline = document.querySelector(
                "[data-testid='visual-builder__hover-outline']"
            );
            expect(hoverOutline).toHaveAttribute("style");

            const customCursor = document.querySelector(
                `[data-testid="visual-builder__cursor"]`
            );

            expect(customCursor).toHaveAttribute('data-icon', 'link');
            expect(customCursor?.classList.contains("visible")).toBeTruthy();
        });
    });

    describe("link field (multiple)", () => {
        let container: HTMLDivElement;
        let firstLinkField: HTMLAnchorElement;
        let secondLinkField: HTMLAnchorElement;
        let visualBuilder: VisualBuilder;

        beforeEach(() => {
            container = document.createElement("div");
            container.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.link_multiple_"
            );
            container.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleHorizontal());

            firstLinkField = document.createElement("a");
            firstLinkField.setAttribute(
                "data-cslp",
                "all_fields.blt366df6233d9915f5.en-us.link_multiple_.0.href"
            );
            firstLinkField.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleLeft());

            secondLinkField = document.createElement("a");
            secondLinkField.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.link_multiple_.1.href"
            );
            secondLinkField.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleRight());

            container.appendChild(firstLinkField);
            container.appendChild(secondLinkField);
            document.body.appendChild(container);

            visualBuilder = new VisualBuilder();
        });

        afterEach(() => {
            visualBuilder.destroy();
        });

        test("should have outline and custom cursor", async () => {
            container.dispatchEvent(mousemoveEvent);
            await waitForHoverOutline();
            const hoverOutline = document.querySelector(
                "[data-testid='visual-builder__hover-outline']"
            );
            expect(hoverOutline).toHaveAttribute("style");

            const customCursor = document.querySelector(
                `[data-testid="visual-builder__cursor"]`
            );

            expect(customCursor).toHaveAttribute('data-icon', 'link');
            expect(customCursor?.classList.contains("visible")).toBeTruthy();
        });

        test("should have outline and custom cursor on individual instances", async () => {
            firstLinkField.dispatchEvent(mousemoveEvent);
            await waitForHoverOutline();
            const hoverOutline = document.querySelector(
                "[data-testid='visual-builder__hover-outline']"
            );
            expect(hoverOutline).toHaveAttribute("style");

            const customCursor = document.querySelector(
                `[data-testid="visual-builder__cursor"]`
            );

            expect(customCursor).toHaveAttribute('data-icon', 'link');
            expect(customCursor?.classList.contains("visible")).toBeTruthy();
        });
    });
});
