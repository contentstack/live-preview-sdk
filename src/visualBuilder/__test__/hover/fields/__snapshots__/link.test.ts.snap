// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`When an element is hovered in visual builder mode > link field (multiple) > should have custom cursor 1`] = `
<div
  class="go1688149084 visual-builder__cursor visible"
  data-icon="link"
  data-testid="visual-builder__cursor"
>
  <div
    class="visual-builder__cursor-wrapper"
  >
    <div
      class="visual-builder__cursor-pointer go3046085498"
    >
      <svg
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0.0167948 0.602214C-0.0784404 0.246792 0.246794 -0.0784402 0.602216 0.0167948L15.6457 4.04767C16.0914 4.1671 16.1259 4.78625 15.6962 4.95447L8.17161 7.90048C8.04753 7.94906 7.94936 8.04723 7.90078 8.17131L4.95449 15.6962C4.78627 16.1259 4.16712 16.0914 4.04769 15.6457L0.0167948 0.602214Z"
          fill="#5D50BE"
        />
      </svg>
    </div>
    <div
      class="visual-builder__cursor-icon go4261356330"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M15.5747 8.4136C15.7945 8.63311 15.7948 8.98927 15.5753 9.2091L9.21902 15.5747C8.99952 15.7946 8.64336 15.7948 8.42353 15.5753C8.2037 15.3558 8.20344 14.9996 8.42294 14.7798L14.7792 8.41419C14.9987 8.19436 15.3549 8.1941 15.5747 8.4136Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M7.6251 10.0076C7.84477 10.2273 7.84477 10.5835 7.6251 10.8031L4.97197 13.4563C4.23333 14.1949 3.81836 15.1967 3.81836 16.2413C3.81836 17.2859 4.23333 18.2878 4.97197 19.0264C5.71062 19.765 6.71243 20.18 7.75704 20.18C8.27427 20.18 8.78644 20.0781 9.2643 19.8802C9.74216 19.6823 10.1764 19.3921 10.5421 19.0264L13.1952 16.3733C13.4149 16.1536 13.7711 16.1536 13.9907 16.3733C14.2104 16.5929 14.2104 16.9491 13.9907 17.1688L11.3376 19.8219C10.8674 20.2921 10.3092 20.6651 9.69482 20.9196C9.08047 21.174 8.42201 21.305 7.75704 21.305C6.41406 21.305 5.1261 20.7715 4.17648 19.8219C3.22685 18.8723 2.69336 17.5843 2.69336 16.2413C2.69336 14.8984 3.22685 13.6104 4.17648 12.6608L6.8296 10.0076C7.04927 9.78798 7.40543 9.78798 7.6251 10.0076Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M12.6608 4.17648C13.6104 3.22685 14.8984 2.69336 16.2413 2.69336C17.5843 2.69336 18.8723 3.22685 19.8219 4.17648C20.7715 5.1261 21.305 6.41406 21.305 7.75704C21.305 9.10001 20.7715 10.388 19.8219 11.3376L17.1688 13.9907C16.9491 14.2104 16.5929 14.2104 16.3733 13.9907C16.1536 13.7711 16.1536 13.4149 16.3733 13.1952L19.0264 10.5421C19.765 9.80345 20.18 8.80164 20.18 7.75704C20.18 6.71243 19.765 5.71062 19.0264 4.97197C18.2878 4.23333 17.2859 3.81836 16.2413 3.81836C15.1967 3.81836 14.1949 4.23333 13.4563 4.97197L10.8031 7.6251C10.5835 7.84477 10.2273 7.84477 10.0076 7.6251C9.78798 7.40543 9.78798 7.04927 10.0076 6.8296L12.6608 4.17648Z"
          fill="white"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`When an element is hovered in visual builder mode > link field (multiple) > should have custom cursor on individual instances 1`] = `
<div
  class="go1688149084 visual-builder__cursor visible"
  data-icon="link"
  data-testid="visual-builder__cursor"
>
  <div
    class="visual-builder__cursor-wrapper"
  >
    <div
      class="visual-builder__cursor-pointer go3046085498"
    >
      <svg
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0.0167948 0.602214C-0.0784404 0.246792 0.246794 -0.0784402 0.602216 0.0167948L15.6457 4.04767C16.0914 4.1671 16.1259 4.78625 15.6962 4.95447L8.17161 7.90048C8.04753 7.94906 7.94936 8.04723 7.90078 8.17131L4.95449 15.6962C4.78627 16.1259 4.16712 16.0914 4.04769 15.6457L0.0167948 0.602214Z"
          fill="#5D50BE"
        />
      </svg>
    </div>
    <div
      class="visual-builder__cursor-icon go4261356330"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M15.5747 8.4136C15.7945 8.63311 15.7948 8.98927 15.5753 9.2091L9.21902 15.5747C8.99952 15.7946 8.64336 15.7948 8.42353 15.5753C8.2037 15.3558 8.20344 14.9996 8.42294 14.7798L14.7792 8.41419C14.9987 8.19436 15.3549 8.1941 15.5747 8.4136Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M7.6251 10.0076C7.84477 10.2273 7.84477 10.5835 7.6251 10.8031L4.97197 13.4563C4.23333 14.1949 3.81836 15.1967 3.81836 16.2413C3.81836 17.2859 4.23333 18.2878 4.97197 19.0264C5.71062 19.765 6.71243 20.18 7.75704 20.18C8.27427 20.18 8.78644 20.0781 9.2643 19.8802C9.74216 19.6823 10.1764 19.3921 10.5421 19.0264L13.1952 16.3733C13.4149 16.1536 13.7711 16.1536 13.9907 16.3733C14.2104 16.5929 14.2104 16.9491 13.9907 17.1688L11.3376 19.8219C10.8674 20.2921 10.3092 20.6651 9.69482 20.9196C9.08047 21.174 8.42201 21.305 7.75704 21.305C6.41406 21.305 5.1261 20.7715 4.17648 19.8219C3.22685 18.8723 2.69336 17.5843 2.69336 16.2413C2.69336 14.8984 3.22685 13.6104 4.17648 12.6608L6.8296 10.0076C7.04927 9.78798 7.40543 9.78798 7.6251 10.0076Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M12.6608 4.17648C13.6104 3.22685 14.8984 2.69336 16.2413 2.69336C17.5843 2.69336 18.8723 3.22685 19.8219 4.17648C20.7715 5.1261 21.305 6.41406 21.305 7.75704C21.305 9.10001 20.7715 10.388 19.8219 11.3376L17.1688 13.9907C16.9491 14.2104 16.5929 14.2104 16.3733 13.9907C16.1536 13.7711 16.1536 13.4149 16.3733 13.1952L19.0264 10.5421C19.765 9.80345 20.18 8.80164 20.18 7.75704C20.18 6.71243 19.765 5.71062 19.0264 4.97197C18.2878 4.23333 17.2859 3.81836 16.2413 3.81836C15.1967 3.81836 14.1949 4.23333 13.4563 4.97197L10.8031 7.6251C10.5835 7.84477 10.2273 7.84477 10.0076 7.6251C9.78798 7.40543 9.78798 7.04927 10.0076 6.8296L12.6608 4.17648Z"
          fill="white"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`When an element is hovered in visual builder mode > link field (multiple) > should have outline 1`] = `
<div
  data-cslp="all_fields.bltapikey.en-us.link_multiple_"
>
  <a
    data-cslp="all_fields.blt366df6233d9915f5.en-us.link_multiple_.0.href"
  />
  <a
    data-cslp="all_fields.bltapikey.en-us.link_multiple_.1.href"
  />
</div>
`;

exports[`When an element is hovered in visual builder mode > link field (multiple) > should have outline 2`] = `
<div
  class="visual-builder__hover-outline visual-builder__hover-outline--unclickable go3177626044 go1719569523"
  data-testid="visual-builder__hover-outline"
  style="top: 34px; left: 34px; width: 828px; height: 54.3984375px;"
/>
`;

exports[`When an element is hovered in visual builder mode > link field (multiple) > should have outline on individual instances 1`] = `
<a
  data-cslp="all_fields.blt366df6233d9915f5.en-us.link_multiple_.0.href"
/>
`;

exports[`When an element is hovered in visual builder mode > link field (multiple) > should have outline on individual instances 2`] = `
<div
  class="visual-builder__hover-outline visual-builder__hover-outline--unclickable go3177626044 go1719569523"
  data-testid="visual-builder__hover-outline"
  style="top: 51px; left: 51px; width: 27.7734375px; height: 20.3984375px;"
/>
`;

exports[`When an element is hovered in visual builder mode > link field > should have custom cursor 1`] = `
<div
  class="go1688149084 visual-builder__cursor visible"
  data-icon="link"
  data-testid="visual-builder__cursor"
>
  <div
    class="visual-builder__cursor-wrapper"
  >
    <div
      class="visual-builder__cursor-pointer go3046085498"
    >
      <svg
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0.0167948 0.602214C-0.0784404 0.246792 0.246794 -0.0784402 0.602216 0.0167948L15.6457 4.04767C16.0914 4.1671 16.1259 4.78625 15.6962 4.95447L8.17161 7.90048C8.04753 7.94906 7.94936 8.04723 7.90078 8.17131L4.95449 15.6962C4.78627 16.1259 4.16712 16.0914 4.04769 15.6457L0.0167948 0.602214Z"
          fill="#5D50BE"
        />
      </svg>
    </div>
    <div
      class="visual-builder__cursor-icon go4261356330"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M15.5747 8.4136C15.7945 8.63311 15.7948 8.98927 15.5753 9.2091L9.21902 15.5747C8.99952 15.7946 8.64336 15.7948 8.42353 15.5753C8.2037 15.3558 8.20344 14.9996 8.42294 14.7798L14.7792 8.41419C14.9987 8.19436 15.3549 8.1941 15.5747 8.4136Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M7.6251 10.0076C7.84477 10.2273 7.84477 10.5835 7.6251 10.8031L4.97197 13.4563C4.23333 14.1949 3.81836 15.1967 3.81836 16.2413C3.81836 17.2859 4.23333 18.2878 4.97197 19.0264C5.71062 19.765 6.71243 20.18 7.75704 20.18C8.27427 20.18 8.78644 20.0781 9.2643 19.8802C9.74216 19.6823 10.1764 19.3921 10.5421 19.0264L13.1952 16.3733C13.4149 16.1536 13.7711 16.1536 13.9907 16.3733C14.2104 16.5929 14.2104 16.9491 13.9907 17.1688L11.3376 19.8219C10.8674 20.2921 10.3092 20.6651 9.69482 20.9196C9.08047 21.174 8.42201 21.305 7.75704 21.305C6.41406 21.305 5.1261 20.7715 4.17648 19.8219C3.22685 18.8723 2.69336 17.5843 2.69336 16.2413C2.69336 14.8984 3.22685 13.6104 4.17648 12.6608L6.8296 10.0076C7.04927 9.78798 7.40543 9.78798 7.6251 10.0076Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M12.6608 4.17648C13.6104 3.22685 14.8984 2.69336 16.2413 2.69336C17.5843 2.69336 18.8723 3.22685 19.8219 4.17648C20.7715 5.1261 21.305 6.41406 21.305 7.75704C21.305 9.10001 20.7715 10.388 19.8219 11.3376L17.1688 13.9907C16.9491 14.2104 16.5929 14.2104 16.3733 13.9907C16.1536 13.7711 16.1536 13.4149 16.3733 13.1952L19.0264 10.5421C19.765 9.80345 20.18 8.80164 20.18 7.75704C20.18 6.71243 19.765 5.71062 19.0264 4.97197C18.2878 4.23333 17.2859 3.81836 16.2413 3.81836C15.1967 3.81836 14.1949 4.23333 13.4563 4.97197L10.8031 7.6251C10.5835 7.84477 10.2273 7.84477 10.0076 7.6251C9.78798 7.40543 9.78798 7.04927 10.0076 6.8296L12.6608 4.17648Z"
          fill="white"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`When an element is hovered in visual builder mode > link field > should have outline 1`] = `
<a
  data-cslp="all_fields.bltapikey.en-us.link.href"
/>
`;

exports[`When an element is hovered in visual builder mode > link field > should have outline 2`] = `
<div
  class="visual-builder__hover-outline visual-builder__hover-outline--unclickable go3177626044 go1719569523"
  data-testid="visual-builder__hover-outline"
  style="top: 51px; left: 51px; width: 27.7734375px; height: 20.3984375px;"
/>
`;
